Add plate 2 feature processing module

jenna-tomkinson opened the pull request:
In this PR, it includes the notebooks/py files to:

1) Merge and convert single cells from CellProfiler SQLite file to parquet file + add single cell counts per well as metadata column
2) Annotate the merged parquet file with metadata from platemap file
3) Normalize features with negative control
4) Feature select with normalized data

roshankern commented on interstellar_data_env.yml:
```diff
@@ -28,4 +28,4 @@ dependencies:
     - git+https://github.com/broadinstitute/pe2loaddata.git@0426dd6b9d8b3242294a6fbdef7c4e1ee762a4cc
     - git+https://github.com/cytomining/pycytominer@7e571322d211f66e716bf135c8b52898556fb960
-    
+    - git+https://github.com/cytomining/CytoTable@e3773183b8fd930a6513c65dd3e45418e9a5e80d
```
The version control here is immaculate with these hashes 👌

jenna-tomkinson replied:
thank you!

roshankern commented on Plate_2_data/4.processing_features/scripts/annotate_sc_plate2.py:
```diff
@@ -0,0 +1,90 @@
+# In[1]:
+
+
+import pandas as pd
+import pathlib
+
+from pycytominer import annotate
+from pycytominer.cyto_utils import output
+
+import sys
+sys.path.append("../../utils")
+import sc_extraction_utils as sc_utils
```
```suggestion
import sys
import pathlib

import pandas as pd
from pycytominer import annotate
from pycytominer.cyto_utils import output

sys.path.append("../../utils")
import sc_extraction_utils as sc_utils
```

Eric has been stressing following the PEP conventions for imports in my PRs and I think we as a lab should start focusing on this more. The [PEP documentation](https://peps.python.org/pep-0008/#imports) says imports should be in the following order:

1. Standard library imports.
2. Related third party imports.
3. Local application/library specific imports.

roshankern replied:
Also, consider this comment for the rest of the python files in this PR and elsewhere!

jenna-tomkinson replied:
Huh, I haven't seen this! I agree that the lab should follow the same conventions so I will make this change for all notebooks/py files!

jenna-tomkinson replied:
See [45c5ceb](https://github.com/WayScience/pyroptosis_signature/pull/27/commits/45c5ceb3358d2427e9742a1a0701bbcadf78fc49) for change.

roshankern commented on Plate_2_data/4.processing_features/scripts/feature_select_sc_plate2.py:
```diff
@@ -0,0 +1,74 @@
+# In[3]:
+
+
+# list of operations for feature select function to use on input profile
+feature_select_ops = [
+    "variance_threshold",
+    "correlation_threshold",
+    "blocklist",
+]
```
Consider adding a small comment as to why these feature select functions were chosen.

jenna-tomkinson replied:
This is a good comment, I actually haven't gone deeper into why these operations are chosen other than these are the standard ones I have seen in pipelines with CellProfiler feature data.

I will start investigating this and add this into another PR. Thank you for bringing this up!

roshankern reviewed (approved):
Great job here, LGTM!

Just a couple of small comments you can address or ignore.

jenna-tomkinson commented on Plate_2_data/4.processing_features/README.md:
````diff
@@ -0,0 +1,21 @@
+Using the code below, execute the `sh` file and merge single cells to use for annotation, normalization, and feature selection.
+
+```bash
+source processing_features_plate2.sh
````
I noticed that when I tried using `bash` when running on my Macbook, it was causing errors, so my hope was to increase reproducibility by changing from `bash` to `source`. 

I will try to go back and make this change in all areas but it will have to be something in a refactor in the future.

Thank you for asking!

jenna-tomkinson commented on Plate_2_data/4.processing_features/processing_features_plate2.sh:
```diff
@@ -0,0 +1,12 @@
+conda activate interstellar_data
+
+# convert all notebooks to python files into the scripts folder
+jupyter nbconvert --to python --output-dir=scripts/ *.ipynb
```
I think since this file in the directory with the four notebooks, then I would rather have one line that converts all of the notebook at once. It makes it look cleaner to me but I can see why `*` could not be preferred depending on the situation.

jenna-tomkinson commented on Plate_2_data/4.processing_features/processing_features_plate2.sh:
```diff
@@ -0,0 +1,12 @@
+jupyter nbconvert --to python --output-dir=scripts/ *.ipynb
+
+# run the python scripts in order (from convert+merge, annotate, normalize, and feature select)
+python scripts/merge_sc_plate2.py && python scripts/annotate_sc_plate2.py && python scripts/normalize_sc_plate2.py && python scripts/feature_select_sc_plate2.py
```
It would work, but I was hoping to continue with the one line command pattern.

This is definitely a preference as I use that command above for my other modules since I normally only have one notebook. I think I will stick with this but I appreciate the perspective on readability since it is a pretty long line.

MikeLippincott commented on Plate_2_data/4.processing_features/scripts/merge_sc_plate2.py:
```diff
@@ -0,0 +1,89 @@
+                "../3.cellprofiler_analysis/analysis_output/SHSY5Y_cells_incomplete_first_run.sqlite"
+            )
+        ),
+        "dest_path": str(pathlib.Path(f"{output_dir}/SHSY5Y_first_run.parquet")),
```
looks like this needs `black` formatter

jenna-tomkinson replied:
Got it! 👍

MikeLippincott commented on Plate_2_data/4.processing_features/scripts/normalize_sc_plate2.py:
```diff
@@ -0,0 +1,75 @@
+    print(f"Normalizing annotated merged single cells for {SHSY5Y_run}!")
+
+    # normalize annotated data
+    normalized_df = normalize(
```
consider adding inline comments as to why/what the args to normalize are so one doesn't need to search for it in a helper file

jenna-tomkinson replied:
Added! 👍

MikeLippincott commented:
LGTM! Minor comments, some doing with future implementations. Cheers! Excited to looks at this data.

jenna-tomkinson commented on Plate_2_data/4.processing_features/processing_features_plate2.sh:
```diff
@@ -0,0 +1,12 @@
+jupyter nbconvert --to python --output-dir=scripts/ *.ipynb
+
+# run the python scripts in order (from convert+merge, annotate, normalize, and feature select)
+python scripts/merge_sc_plate2.py && python scripts/annotate_sc_plate2.py && python scripts/normalize_sc_plate2.py && python scripts/feature_select_sc_plate2.py
```
I think that the print statements in the terminal will work for now, but I will keep this as a consideration. Thank you!

jenna-tomkinson commented:
@roshankern @MikeLippincott 

Thank you both for reviewing this PR. I am going to merge this now after addressing the comments. This will be the last PR for a while in this repo, as we will be working on Alpine implementation to run the PBMCs (TBD).
